chain/deals/client_states.go

```diff
@@ -1,142 +1,224 @@
 package deals
 
 import (
+	"bytes"
 	"context"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/lib/cborutil"
 
 	"golang.org/x/xerrors"
 
 	"github.com/filecoin-project/lotus/api"
 	"github.com/filecoin-project/lotus/chain/actors"
 	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/types"
 )
 
-type clientHandlerFunc func(ctx context.Context, deal ClientDeal) error
+type clientHandlerFunc func(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error)
 
 func (c *Client) handle(ctx context.Context, deal ClientDeal, cb clientHandlerFunc, next api.DealState) {
 	go func() {
-		err := cb(ctx, deal)
+		mut, err := cb(ctx, deal)
 		if err != nil {
 			next = api.DealError
 		}
+
+		if err == nil && next == api.DealNoUpdate {
+			return
+		}
+
 		select {
 		case c.updated <- clientDealUpdate{
 			newState: next,
 			id:       deal.ProposalCid,
 			err:      err,
+			mut:      mut,
 		}:
 		case <-c.stop:
 		}
 	}()
 }
 
-func (c *Client) new(ctx context.Context, deal ClientDeal) error {
+func (c *Client) new(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error) {
 	resp, err := c.readStorageDealResp(deal)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	if err := c.disconnect(deal); err != nil {
+		return nil, err
 	}
 
+	/* data transfer happens */
 	if resp.State != api.DealAccepted {
-		return xerrors.Errorf("deal wasn't accepted (State=%d)", resp.State)
+		return nil, xerrors.Errorf("deal wasn't accepted (State=%d)", resp.State)
 	}
 
-	// TODO: spec says it's optional
-	pubmsg, err := c.chain.GetMessage(*resp.PublishMessage)
+	return func(info *ClientDeal) {
+		info.PublishMessage = resp.PublishMessage
+	}, nil
+}
+
+func (c *Client) accepted(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error) {
+	log.Infow("DEAL ACCEPTED!")
+
+	pubmsg, err := c.chain.GetMessage(*deal.PublishMessage)
```

Comment:
We might see this one failing with some frequency, the message isnt likely to propagate faster than the miner can send us an update directly

Reply:
We can put the message in the deal response the miner sends us. Does that make sense?

Reply:
Yeah, that would be good I think

Reply:
We could also just do the wait first, before this validation.